design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,325 @@
+{
+"query": "select * from foo", <----- the SQL of the query to execute
+"push": true,                 <----- if true then push query, else pull query
+"limit": 10                   <----  If specified return at most this many rows,
+"properties": {               <----- Optional properties for the query
+    "prop1": "val1",
+    "prop2": "val2"
+   }
+}
+
+````
+
+Please note the parameters are not necessarily exhaustive. The description here is an outline not a detailed
+low level design. The low level design will evolve during development.
+
+In the case of a successful query 
+
+````
+{
+"query-id", "xyz123",                          <---- unique ID of the query, used when terminating the query
+"columns":["col", "col2", "col3"],             <---- the names of the columns
+"column_types":["BIGINT", "STRING", "BOOLEAN"] <---- The types of the columns
+"row_count": 101                               <---- The number of rows - only set in case of pull query
+}
+````
+
+Followed by zero or more JSON arrays:
+
+````
+[123, "blah", true]
+[432, "foo", true]
+[765, "whatever", false]
+````
+
+Each JSON array or row will be delimited by a newline.
+
+For a pull query the response will be ended by the server once all rows have been written, for
+a push query the response will remain open until the connection is closed or the query is explicitly
+terminated.
+
+#### Terminating queries
+
+Push queries can be explicitly terminated by the client by making a request to this endpoint
+
+The request method will be a POST.
+
+Requests will be made to a specific URL, e.g. "/query-terminate" (this can be configurable)
+
+The body of the request is a JSON object UTF-8 encoded as text, containing the arguments for the
+operation (newlines have been added here for the sake of clarity but the real JSON must not contain newlines)
+
+````
+{
+"query-id": "xyz123", <----- the ID of the query to terminate
+}
+
+````
+ 
+
+#### Inserts
+
+The request method will be a POST.
+
+Requests will be made to a specific URL, e.g. "/insert-stream" (this can be configurable)
+
+The body of the request is a JSON object UTF-8 encoded as text, containing the arguments for the
+operation (newlines have been added for clarity, the real JSON must not contain newlines):
+
+````
+{
+"stream": "my-stream" <----- The name of the KSQL stream to insert into
+"acks": true          <----- If true then a stream of acks will be returned in the response
+}
+
+````
+
+Followed by zero or more JSON objects representing the values to insert:
+
+````
+{
+"col1" : "val1",
+"col2": 2.3,
+"col3", true
+}
+````
+Each JSON object will be separated by a new line.
+
+To terminate the insert stream the client should end the request.
+
+If acks are requested then the response will be written to the response when each row has been
+successfully committed to the underlying topic. Rows are committed in the order they are provided.
+Each ack in the response is just an empty JSON object, separated by newlines:
```

Comment:
I too have concerns around the error path. 

If we can stream results to the server, and get an empty ack message streamed back on success, then I'm wondering what we're expecting to happen on failure?

Are we going to end up with all client impls needing to track in-flight requests?

To allow clients to correlate error response to request we could always include the request itself as part of the error response.

Also, when is the ACK sent back?  Once the server has processed the request and enqued the message to be produced to Kafka, or when the producer call back confirms the message was produced to Kafka?  Hopefully the second.

Reply:
>> Should the KLIP not reflect the exact definition? IMHO, the KLIP should allow anybody to write a ksqlDB client in their language of choice.

There will be API documentation on the web-site which will explain the exact endpoints, arguments and how to use them.

Reply:
>> What happens is a record cannot be inserted for whatever reason (for example a single record could exceed the maximum allowed message size, if the JSON is invalid and cannot be parsed).

In case of error, an error response is sent back on the same stream and the stream is closed.

Reply:
>> Hopefully the second.

Ack!

Reply:
>> To allow clients to correlate error response to request we could always include the request itself as part of the error response.

This is unnecessary as acks are always returned in the same order the inserts were sent. From the clients point of view this makes correlation of responses a bit simpler as they don't have to maintain a map of insert_id, request. They just need to store a queue of requests and when responses come back pop the head of the queue.

Reply:
Ok...on second thoughts, I am going to send back the send_id in the ack. The reason being that on the server, acks for sends to Kafka can come back in a different order if they are sent to different partitions. I don't want to have to sort the acks on the server, so I'll send them back as they come in and the client can correlate them.